tools/pd-ctl/pdctl/command/global.go

```diff
@@ -55,69 +55,128 @@ func InitHTTPSClient(CAPath, CertPath, KeyPath string) error {
+		}
+		// the resp would be returned by the outer function
+		resp, err = dail(req)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	return resp, err
 }
 
 func dail(req *http.Request) (string, error) {
-	var res string
-	reps, err := dialClient.Do(req)
+	resp, err := dialClient.Do(req)
 	if err != nil {
-		return res, err
-	}
-	defer reps.Body.Close()
-	if reps.StatusCode != http.StatusOK {
-		return res, genResponseError(reps)
+		return "", err
 	}
-
-	r, err := ioutil.ReadAll(reps.Body)
-	if err != nil {
-		return res, err
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		var msg []byte
+		msg, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		return fmt.Sprintf("[%d] %s", resp.StatusCode, msg), nil
 	}
-	res = string(r)
-	return res, nil
-}
 
-func doRequest(cmd *cobra.Command, prefix string, method string) (string, error) {
-	req, err := getRequest(cmd, prefix, method, "", nil)
+	content, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
 		return "", err
 	}
-	return dail(req)
+	return string(content), nil
 }
 
-func genResponseError(r *http.Response) error {
-	res, _ := ioutil.ReadAll(r.Body)
-	return errors.Errorf("[%d] %s", r.StatusCode, res)
-}
+// DoFunc receives an endpoint which you can issue request to
+type DoFunc func(endpoint string) error
 
-func getAddressFromCmd(cmd *cobra.Command, prefix string) string {
-	p, err := cmd.Flags().GetString("pd")
+// roundRobinURLs returns a closure which runs a DoFunc for each
+// url parsed from the command line
+func roundRobinURLs(cmd *cobra.Command) func(f DoFunc) {
+	addrs, err := cmd.Flags().GetString("pd")
 	if err != nil {
-		cmd.Println("get pd address error, should set flag with '-u'")
+		cmd.Println("get pd address failed, should set flag with '-u'")
 		os.Exit(1)
 	}
+	endpoints := strings.Split(addrs, ",")
+	for i, endpoint := range endpoints {
+		u, err := url.Parse(endpoint)
+		if err != nil {
+			cmd.Println("address format is wrong, should like 'http://127.0.0.1:2379' or '127.0.0.1:2379'")
+			os.Exit(1)
+		}
+		// tolerate some schemes that will be used by users, the TiKV SDK
+		// use 'tikv' as the scheme, it is really confused if we do not
+		// support it by pdctl
+		if u.Scheme == "" || u.Scheme == "pd" || u.Scheme == "tikv" {
+			u.Scheme = "http"
+		}
 
-	if p != "" && !strings.HasPrefix(p, "http") {
-		p = "http://" + p
+		endpoints[i] = u.String()
 	}
 
-	u, err := url.Parse(p)
-	if err != nil {
-		cmd.Println("address format is wrong, should like 'http://127.0.0.1:2379' or '127.0.0.1:2379'")
+	total := len(endpoints)
+	next := 0
+	return func(f DoFunc) {
+		first := next
+		for {
+			endpoint := endpoints[next]
+			// when there is an error, try next endpoint, keep using
+			// the health endpoint otherwise.
+			if err := f(endpoint); err != nil {
+				next = (next + 1) % total
+				// we have tried all the endpoints, now print the error
+				// and break
+				if first == next {
+					cmd.Println(err)
```

Comment:
If we have tried all the endpoints, the error message here may be confused?

Reply:
Why is it confusing, and what do you suggest?